hare/broker.go

```diff
@@ -77,49 +84,61 @@ func (broker *Broker) Start() error {
 	return nil
 }
 
-type Message struct {
-	msg            *pb.HareMessage
-	bytes          []byte
-	validationChan chan service.MessageValidation
-}
-
-func (msg Message) reportValidationResult(isValid bool) {
-	if msg.validationChan != nil {
-		msg.validationChan <- service.NewMessageValidation(msg.bytes, ProtoName, isValid)
-	}
-}
-
 // Dispatch incoming messages to the matching set id instance
 func (broker *Broker) dispatcher() {
 	for {
 		select {
 		case msg := <-broker.inbox:
+			if msg == nil {
+				log.Warning("Message validation failed: called with nil")
+				msg.ReportValidation(ProtoName, false)
+				continue
+			}
+
 			hareMsg := &pb.HareMessage{}
 			err := proto.Unmarshal(msg.Bytes(), hareMsg)
 			if err != nil {
 				log.Error("Could not unmarshal message: ", err)
 				continue
 			}
 
-			instanceId := NewBytes32(hareMsg.Message.InstanceId)
+			// message validation
+			if hareMsg.Message == nil {
+				log.Warning("Message validation failed: message is nil")
+				msg.ReportValidation(ProtoName, false)
+				continue
+			}
+
+			if hareMsg.Message.InstanceId > broker.maxReg+1 { // intended for future unregistered instance
+				log.Warning("Message validation failed: instanceId. Max: %v Actual: %v", broker.maxReg, hareMsg.Message.InstanceId)
+				msg.ReportValidation(ProtoName, false)
+				continue
+			}
+
+			if !broker.eValidator.Validate(hareMsg) {
+				log.Warning("Message validation failed: eValidator returned false %v", hareMsg)
+				msg.ReportValidation(ProtoName, false)
+				continue
+			}
+
+			// validation passed
+			msg.ReportValidation(ProtoName, true)
 
+			instanceId := InstanceId(hareMsg.Message.InstanceId)
 			broker.mutex.RLock()
 			c, exist := broker.outbox[instanceId.Id()]
 			broker.mutex.RUnlock()
-			mOut := Message{hareMsg, msg.Bytes(), msg.ValidationCompletedChan()}
 			if exist {
 				// todo: err if chan is full (len)
-				c <- mOut
+				c <- hareMsg
 			} else {
 
 				broker.mutex.Lock()
 				if _, exist := broker.pending[instanceId.Id()]; !exist {
-					broker.pending[instanceId.Id()] = make([]Message, 0)
+					broker.pending[instanceId.Id()] = make([]*pb.HareMessage, 0)
 				}
-				broker.pending[instanceId.Id()] = append(broker.pending[instanceId.Id()], mOut)
+				broker.pending[instanceId.Id()] = append(broker.pending[instanceId.Id()], hareMsg)
```

Comment:
are there any concurrent writes to pending[] map? should it be locked if so? I'd recommend running your unit tests with --race before you push

Reply:
I am protecting it with the lock. Let me know if you see a scenario in which I do not. Didn't see issues with -race regarding this test.

Reply:
I am not sure it is a good idea because I will eventually want to remove the inboxer interface and return the inbox. That will complicate the register call. We can probably solve it by passing a function on the channel but it sounds too complicated.

Reply:
you can still make this change and return the channel. Just create the channel and pass it to the dispatcher via the register channel. You can make this change as well as removing the inboxer as part of this PR.